format/leveldb/leveldb_ldb.go

```diff
@@ -0,0 +1,265 @@
+		d.FieldU64("magic_number", d.UintAssert(tableMagicNumber), scalar.UintHex)
+	})
+
+	// metaindex
+
+	d.SeekAbs(metaIndexOffset * 8)
+	var metaHandles []BlockHandle
+	readBlock("metaindex", metaIndexSize, readKeyValueContents, func(d *decode.D) {
+		// BlockHandle
+		// https://github.com/google/leveldb/blob/main/table/format.cc#L24
+		handle := BlockHandle{
+			Offset: d.FieldULEB128("offset"),
+			Size:   d.FieldULEB128("size"),
+		}
+		metaHandles = append(metaHandles, handle)
+	}, d)
+
+	// index
+
+	d.SeekAbs(indexOffset * 8)
+	var dataHandles []BlockHandle
+	readBlock("index", indexSize, readKeyValueContents, func(d *decode.D) {
+		// BlockHandle
+		// https://github.com/google/leveldb/blob/main/table/format.cc#L24
+		handle := BlockHandle{
+			Offset: d.FieldULEB128("offset"),
+			Size:   d.FieldULEB128("size"),
+		}
+		dataHandles = append(dataHandles, handle)
+	}, d)
+
+	// meta
+
+	if len(metaHandles) > 0 {
+		d.FieldArray("meta", func(d *decode.D) {
+			for _, handle := range metaHandles {
+				d.SeekAbs(int64(handle.Offset) * 8)
+				readBlock("meta_block", int64(handle.Size), readMetaContent, nil, d)
+			}
+		})
+	}
+
+	// data
+
+	if len(dataHandles) > 0 {
+		d.FieldArray("data", func(d *decode.D) {
+			for _, handle := range dataHandles {
+				d.SeekAbs(int64(handle.Offset) * 8)
+				readBlock("data_block", int64(handle.Size), readKeyValueContents, nil, d)
+			}
+		})
+	}
+
+	return nil
+}
+
+// Readers
+
+// Read block contents as well as compression + crc bytes following it.
+// The function `readBlockContents` gets the uncompressed bytebuffer.
+// https://github.com/google/leveldb/blob/main/table/format.cc#L69
+func readBlock(name string, size int64, readBlockContents func(size int64, valueCallbackFn func(d *decode.D), d *decode.D), valueCallbackFn func(d *decode.D), d *decode.D) {
+	d.FieldStruct(name, func(d *decode.D) {
+		start := d.Pos()
+		br := d.RawLen(size * 8)
+		// compression (1 byte)
+		compressionType := d.FieldU8("compression", compressionTypes, scalar.UintHex)
+		// crc (4 bytes)
+		data := d.ReadAllBits(br)
+		bytesToCheck := append(data, uint8(compressionType))
+		maskedCRCInt := maskedCrc32(bytesToCheck)
+		d.FieldU32("crc", d.UintAssert(uint64(maskedCRCInt)), scalar.UintHex)
+		// decompress if needed
+		d.SeekAbs(start)
+		if compressionType == compressionTypeNone {
+			d.FieldStruct("uncompressed", func(d *decode.D) {
+				readBlockContents(size, valueCallbackFn, d)
+			})
+		} else {
+			compressedSize := size
+			compressed := data
+			bb := &bytes.Buffer{}
+			switch compressionType {
+			case compressionTypeSnappy:
+				decompressed, err := snappy.Decode(nil, compressed)
+				if err != nil {
+					d.Fatalf("failed decompressing data: %v", err)
+				}
+				d.Copy(bb, bytes.NewReader(decompressed))
+			default:
+				d.Fatalf("Unsupported compression type: %x", compressionType)
+			}
+			d.FieldStructRootBitBufFn("uncompressed", bitio.NewBitReader(bb.Bytes(), -1), func(d *decode.D) {
```

Comment:
I really should come up with a nicer API for the nested decoding stuff, feels a bit like a hack atm